app/module/module.go

```diff
@@ -0,0 +1,349 @@
+package module
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	sdkmodule "github.com/cosmos/cosmos-sdk/types/module"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// Manager defines a module manager that provides the high level utility for managing and executing
+// operations for a group of modules
+type Manager struct {
+	versionedModules   map[uint64]map[string]sdkmodule.AppModule
+	allModules         []sdkmodule.AppModule
+	firstVersion       uint64
+	lastVersion        uint64
+	OrderInitGenesis   []string
+	OrderExportGenesis []string
+	OrderBeginBlockers []string
+	OrderEndBlockers   []string
+	OrderMigrations    []string
+}
+
+type VersionedModule struct {
+	module                 sdkmodule.AppModule
+	fromVersion, toVersion uint64
+}
+
+func NewVersionedModule(module sdkmodule.AppModule, fromVersion, toVersion uint64) VersionedModule {
+	return VersionedModule{
+		module:      module,
+		fromVersion: fromVersion,
+		toVersion:   toVersion,
+	}
+}
+
+// NewManager creates a new Manager object
+func NewManager(modules ...VersionedModule) (*Manager, error) {
+	moduleMap := make(map[uint64]map[string]sdkmodule.AppModule)
+	allModules := make([]sdkmodule.AppModule, len(modules))
+	modulesStr := make([]string, 0, len(modules))
+	firstVersion, lastVersion := uint64(0), uint64(0)
+	for idx, module := range modules {
+		if module.fromVersion == 0 {
+			return nil, sdkerrors.ErrInvalidVersion.Wrapf("v0 is not a valid version for module %s", module.module.Name())
+		}
+		if module.fromVersion > module.toVersion {
+			return nil, sdkerrors.ErrLogic.Wrapf("toVersion can not be less than fromVersion for module %s", module.module.Name())
+		}
+		for version := module.fromVersion; version <= module.toVersion; version++ {
+			if moduleMap[version] == nil {
+				moduleMap[version] = make(map[string]sdkmodule.AppModule)
+			}
+			moduleMap[version][module.module.Name()] = module.module
+		}
+		allModules[idx] = module.module
+		modulesStr = append(modulesStr, module.module.Name())
+		if firstVersion == 0 || module.fromVersion < firstVersion {
+			firstVersion = module.fromVersion
+		}
+		if lastVersion == 0 || module.toVersion > lastVersion {
+			lastVersion = module.toVersion
+		}
+	}
+
+	return &Manager{
+		versionedModules:   moduleMap,
+		allModules:         allModules,
+		firstVersion:       firstVersion,
+		lastVersion:        lastVersion,
+		OrderInitGenesis:   modulesStr,
+		OrderExportGenesis: modulesStr,
+		OrderBeginBlockers: modulesStr,
+		OrderEndBlockers:   modulesStr,
```

Comment:
[tangent] I don't think we need to do this at the same time as this PR but I wonder if all these could use the same order: https://github.com/celestiaorg/celestia-app/issues/3052.

Reply:
Yeah I was merely copying across from the SDK